sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Blob/ContainerRegistryBlobClient.cs

```diff
@@ -368,6 +373,98 @@ public virtual async Task<Response<UploadBlobResult>> UploadBlobAsync(Stream str
             }
         }
 
+        private async Task<ChunkedUploadResult> UploadInChunks(string location, Stream stream, int maxChunkSize, bool async = true, CancellationToken cancellationToken = default)
+        {
+            long length = stream.Length;
+
+            string digest = default;
+            ResponseWithHeaders<ContainerRegistryBlobUploadChunkHeaders> uploadChunkResult = null;
+
+            if (length < maxChunkSize)
+            {
+                digest = OciBlobDescriptor.ComputeDigest(stream);
+
+                uploadChunkResult = async ?
+                    await _blobRestClient.UploadChunkAsync(location, stream, cancellationToken: cancellationToken).ConfigureAwait(false) :
+                    _blobRestClient.UploadChunk(location, stream, cancellationToken: cancellationToken);
+
+                return new ChunkedUploadResult(digest,
+                    uploadChunkResult.Headers.Location,
+                    length);
+            }
+
+            byte[] buffer = new byte[maxChunkSize];
+            var ranges = GetRanges(length, maxChunkSize);
+            using SHA256 sha256 = SHA256.Create();
+
+            for (int i = 0; i < ranges.Count; i++)
+            {
+                var range = ranges[i];
+
+                location = uploadChunkResult?.Headers.Location ?? location;
+                var chunkLength = async ?
+                    await stream.ReadAsync(buffer, 0, (int)range.Length, cancellationToken).ConfigureAwait(false) :
+                    stream.Read(buffer, 0, (int)range.Length);
+
+                if (i < ranges.Count - 1)
+                {
+                    sha256.TransformBlock(buffer, 0, chunkLength, buffer, 0);
+                }
+                else
+                {
+                    sha256.TransformFinalBlock(buffer, 0, chunkLength);
+                    digest = "sha256:" + OciBlobDescriptor.BytesToString(sha256.Hash);
+                }
+
+                using (Stream chunk = new MemoryStream(buffer, 0, chunkLength))
+                {
+                    uploadChunkResult = async ?
```

Comment:
>uploadChunkResult [](http://example.com/codeflow?start=20&length=17)

what happens if we fail to upload one of the chunks? Does this rest API call do the retries? 

Also, wouldn't the last uploadChunkResult be used? Is there a chance that the last one may not be the correct representative for all the chunks?

Reply:
The pipeline will automatically handle retries according to the [RetryPolicy](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Pipeline/RetryPolicy.cs).  The caller can configure the RetryPolicy in [RetryOptions](https://apiview.dev/Assemblies/Review/77bdbcef49a3465fa96fac3fb745bf57#Azure.Core.RetryOptions), or starting with Azure.Core 1.26.0, they can implement their own retry policy as well.

>  Is there a chance that the last one may not be the correct representative for all the chunks?

I'm not sure I understand the question ... the API requires that chunks are uploaded [in order](https://docs.docker.com/registry/spec/api/#chunked-upload), and will return a 416 response if an out-of-order chunk is uploaded.  

What do you mean by representative?